### PR TITLE
Revert "feat(UI): Additional zoom levels (#8081)"

### PR DESCRIPTION
**Pull request by warp-core**
For various reasons:

- @quyykk has noted that changing the zoom indices is undesirable.
- The new lower zoom levels result in previously reasonable asteroid densities consuming a significant amount of GPU time, and not drawing the asteroids altogether has been noted as being undesirable due to the interactions asteroids have with missiles (and any projectiles, but missiles specifically since they tend to have the longest ranges).
- Planet label overlapping becomes an issue again.
- Probably other stuff, check the bug reporting channel on Discord iff you want to know more.

### CODE REVIEW

**Review by Zitchas — DISMISSED** [review]
Higher zoom levels were clearly desired by some of us, and have some significant benefits as systems get bigger. As such, instead of just reverting this, I would suggest that we instead make it functional.

In this particular case, I see two potential options:
1. We do fixes to make this mode functional as just another zoom level. (so we do something to make asteroids work, planet labels, etc.
2. We treat this as a different type of visual layer.

My preference would be for 2, because that retains the vastly wider field of view that is the whole point of having the higher zoom levels, and potentially allows us to see the system in a different light that provides more information, or at least presents it differently.

These higher zoom levels could be a tactical view:
- Remove asteroids from display entirely, and instead display some statistical information about the system (asteroid density, solar wind strength, solar power strength).
- Probably hide haze and starscape as well, perhaps replaced with a grey tactical grid of some kind.
- We could show orbits (this also makes it obvious where the center of the system is)
- Replace missile, fighter, and drone sprites with appropriate icons.


In other words, I don't see reversion as being necessary. These additional zoom levels are necessary for the desired functionality, we simply need to take steps to mitigate a couple elements that render it less functional.

**Review by Amazinite — APPROVED** [review]
I don't think we should have a less functional middle-ground. If some want a more tactical view, then that view should be something which is finished and functional before merging, not something we inch toward in such a way as to cause problems for those who don't want this view.

**Comment by Amazinite** [comment]
I think it's also fair to note that #8081 never mentioned anything about such a tactical view, only mentioning #8071, which is unrelated to such a tactical view. I don't find it appropriate to merge a PR under the guise of one issue, which the OP [even states](https://github.com/endless-sky/endless-sky/discussions/8071#discussioncomment-4615774) wouldn't really solve the problem:
> An additional zoom level would also be good, but it wouldn't address the root of the issue. You'd still end up with different levels of zoom depending on your UI scale.

then when we point out that it isn't really a good solution to the stated problem and try to revert it, it be defended on an entirely different ground.

**Comment by Zitchas** [comment]
Sorry, I don't mention every reason I might have for implementing something, including ones that I might happen to think of later. It's a solution that solved the baseline problem of needing to be able to see farther out. Yes, it also happens to provide other benefits, such as being able to see larger patterns, ship movements, and other things farther out. It also makes navigating large systems much nicer, not to mention dealing with the Hai and other groups that have beyond-the-radar weaponry.

In any case, I feel that it is inappropriate to leap straight to "revert that" instead of filing an issue and exploring alternative measures that improve the game instead.

**Review by Hecter94 — APPROVED** [review]
We got some good feedback on this change, and have identified some issues. We can revert for now and look at implementing it again once we've worked out those issues and have a solid idea of how we want to move forward. 

Changes to OSS should be things that we're both quick to merge and quick to revert when needed. We should not be shying away from merging when something looks good, nor should we shy away from reverting a change once it has been playtested and issues identified. 

There is no permanency in OSS. We shouldn't act like a change is permanent as soon as it gets merged.

**Comment by Zitchas** [comment]
I agree to that, quick merging and reverting are both equally valid. What I'm objecting to is the fact that no-one raised an issue about anything, nor has anyone ventured any solutions other than merely reverting it. If there's no problems being listed out, then it's effectively impossible to work towards resolving them.

**Review by Saugia — APPROVED** [review]

**Comment by Terin** [comment]
While there are the (fairly technically involves) steps of mitigation listed above, the point about genre is also pretty important. Ares was brought up as having that kind of tactical layer, at which point that lays out what's going on: Ares and Escape Velocity are two different genres of games. While on paper it's not _that_ much to get a tactical display working, it does start to shift the game in some major ways that aren't in line with how it's been played before.

I didn't realize (and had the time to find out before) that the original PR was zooming out further, just thought it was adding additional steps in-between. So yeah, the performance issues on top of the engine not being built to display at that scale is probably worth a revert at the time.

**Comment by quyykk** [comment]
> By indices, does this mean the index=X entry? Or the list of zooms itself?

The index of the zoom value is stored in preferences, so by changing the list in the middle it changes the index of the preference, and thus the zoom

**Review by quyykk — APPROVED** [review]

**Comment by Zitchas** [comment]
> > By indices, does this mean the index=X entry? Or the list of zooms itself?
> 
> The index of the zoom value is stored in preferences, so by changing the list in the middle it changes the index of the preference, and thus the zoom

OK. Is this not a case where the player will just have to update their preferences? Seems simple enough.

File: source/Engine.cpp [hunk 1]
```diff
@@ -967,40 +967,27 @@ void Engine::Draw() const
...
+		static const Color color[8] = {
+			*colors.Get("overlay friendly shields"),
+			*colors.Get("overlay hostile shields"),
+			*colors.Get("overlay outfit scan"),
+			*colors.Get("overlay friendly hull"),
+			*colors.Get("overlay hostile hull"),
+			*colors.Get("overlay cargo scan"),
+			*colors.Get("overlay friendly disabled"),
+			*colors.Get("overlay hostile disabled")
+		};
```
Comment (RisingLeaf) [inline source/Engine.cpp:981]: can we move this outside the loop?
because micro optimization

**Review by RisingLeaf — CHANGES_REQUESTED** [review]

File: source/Engine.cpp (hunk 1 above)
Comment (quyykk) [reply source/Engine.cpp:981]: wait a second was this array always static? :mad_quyykk:

Comment (RisingLeaf) [reply source/Engine.cpp:981]: oh, right, if its static it is fine :stupid_leaf:

Comment (quyykk) [reply source/Engine.cpp:981]: yeah but it really shouldn't be

Comment (warp-core) [reply source/Engine.cpp:981]: Yes, for reference: https://en.cppreference.com/w/cpp/language/storage_duration#Static_local_variables

Comment (RisingLeaf) [reply source/Engine.cpp:981]: lets make it a const outside then

**Comment by Zitchas** [comment]
> While there are the (fairly technically involves) steps of mitigation listed above, the point about genre is also pretty important. Ares was brought up as having that kind of tactical layer, at which point that lays out what's going on: Ares and Escape Velocity are two different genres of games. While on paper it's not _that_ much to get a tactical display working, it does start to shift the game in some major ways that aren't in line with how it's been played before.
> 
> I didn't realize (and had the time to find out before) that the original PR was zooming out further, just thought it was adding additional steps in-between. So yeah, the performance issues on top of the engine not being built to display at that scale is probably worth a revert at the time.

Regarding the genre - If we want to compare features, ES is already significantly closer to being Ares than EV as is: It has the zoom, it has fleet support, numbered formations, we can click to give "goto orders," give instructions to conserve ammo... These are all stuff that are Ares features and not one's found in Escape Velocity. They're also things that are useless in a deathball or solo close quarters combat; and stuff that facilitates fleet/tactical play.

What my goal/interest is in having a tactical/far out level of zoom is to enable the player to have more of a situational awareness so that they can be more comfortable and capable of engaging in solo combat. I know personally one big push for just using a fleet everywhere is the fact that I can't see what I'm getting into, and stuff will start pummeling me in some areas from beyond radar range, so having other ships along to absorb that is the only way to even the playing field. 

With the greater view range, I can actually see what's coming, and see what I'm getting into it. I can pick my moment to engage, and I can start formulating my tactics and dodging before it's too late to do so. The AI has perfect situational awareness already (and perfect aim, and perfect range calculations), so letting the player see some of this is merely evening the playing field and giving people the chance to make appropriate decisions instead.

It's all part and parcel of my efforts to spread out combat, enable multiple engagement zones and ranges, and make combat more interesting and diversified.

**Comment by RisingLeaf** [comment]
I do not think we should care about what genre we are going for, but just implement things we like.

That said, I think you have a point here that should be considered.

Right now strategic combat in ES is not a big thing (>Deathballing) and I personally would enjoy to have more strategic options in the game. On the other hand I think a smaller view range can also contribute to the game, it especially decreases the knowledge of the current situation.

So while I like to see more strategic setups, I do not know if this is the right thing.

**Comment by Zitchas** [comment]
> Right now strategic combat in ES is not a big thing (>Deathballing) and I personally would enjoy to have more strategic options in the game. On the other hand I think a smaller view range can also contribute to the game, it especially decreases the knowledge of the current situation.

So I guess that's the question, then: Do we intentionally limit the player's field of view such that they can't see situations until they're likely embroiled in them? Or do we allow them a field of view that lets them make smarter choices.

I feel that it is important to note that we have frequently talked about making big ships heavier and thus both slower and less responsive. This will, effectively, make it so that they will require much greater sight-lines to be able to react to anything. 

This is also significant in pointing us towards what sort of combat we want: The shorter the lines of sight that the players have, the more likely they are to feel forced into a "brawl and tank" build. 

There's also the interest factor: Shooting missiles at something off-screen is, at least to me, very boring. It's one of the many reasons I don't use missiles. It's also one reason why I don't send drones & fighters off to harry my target while I'm a long ways away. It's all off-screen action, and thus doesn't have nearly as much excitement or impact. Watching my escort's status bars wiggle up and down on the side of my screen just isn't very fun.

**Review by EjoThims — APPROVED** [review]
I would still love far extended view ranges (yes, in part so we -can- have actually meaningful differences in weapon and engagement ranges instead of the difference between short and long range weapons being smaller than many of our ships).

But it's clear the current way we do things means just deepening the field outright causes unacceptable scaling of resource usage.

Leaving this unacceptable state until fixes are sorted and found is not the way forward; this change needs to be reverted until solutions to the other problems blocking it are ready.

**Comment by Hurleveur** [comment]
We could just have a hidden preference for min and max zoom values and let the game automatically figure out the progression of it

**Comment by Zitchas** [comment]
OK, if we need to remove the farther out zooms, is there reason we can't leave the smoother sequence and simply just remove the farthest out zooms?

**Comment by quyykk** [comment]
> We could just have a hidden preference for min and max zoom values and let the game automatically figure out the progression of it

No, this is what plugins that modify the interface file are for

**Comment by Hurleveur** [comment]
Yeah it makes more sense as part of interface.txt, but you get my point nontheless
